refactor(cs): Encapsulate high level operations

dmga44 opened the pull request:
This change targets encapsulating the high level operations logic and
extract it from within the ChunkserverEntry struct. This is done to
pave the way for more complex changes in the chunkserver operations,
such as write buffering.

The high level operation classes aim to expose a relativaly simple
interface and keep for itself most of the logic of each of those
operations. The callbacks will be directly called upon the HLO
instances instead of some ChunkserverEntry passed via parameters. The
new classes need to access certain members of the ChunkserverEntry
struct and this is restrained in the HighLevelOp class to avoid
missuse.

As a side change, we've remove the extra parameter from the client
bgjobs functions. It is not necessary anymore.

Signed-off-by: Dave <dave@leil.io>

dmga44 committed:
refactor(cs): Encapsulate high level operations

This change targets encapsulating the high level operations logic and
extract it from within the ChunkserverEntry struct. This is done to
pave the way for more complex changes in the chunkserver operations,
such as write buffering.

The high level operation classes aim to expose a relativaly simple
interface and keep for itself most of the logic of each of those
operations. The callbacks will be directly called upon the HLO
instances instead of some ChunkserverEntry passed via parameters. The
new classes need to access certain members of the ChunkserverEntry
struct and this is restrained in the HighLevelOp class to avoid
missuse.

As a side change, we've remove the `extra` parameter from the client
bgjobs functions. It is not necessary anymore.

Signed-off-by: Dave <dave@leil.io>

dmga44 commented:
@gemini-code-assist Would you take a look?

dmga44 commented on src/chunkserver/chunkserver_entry.cc:
```diff
@@ -979,39 +577,32 @@ void ChunkserverEntry::testChunk(const uint8_t *data, uint32_t length) {
+bool ChunkserverEntry::isChunkOpen() {
+	if (writeHLO->isChunkOpen() && readHLO->isChunkOpen()) {
+		safs::log_warn(
+		    "({}) Both write and read chunk handles are open for chunk {:016X} and {:016X}",
+		    __func__, writeHLO->chunkId(), readHLO->chunkId());
 	}
 
```
This should not happen, it is logged, but the chunk needs to be close in any case, so it should be tru for this case too.

dmga44 commented on src/chunkserver/bgjobs.h:
```diff
@@ -277,24 +273,18 @@ uint32_t job_prefetch(JobPool &jobPool, uint64_t chunkId, ChunkPartType chunkTyp
-                        uint64_t chunkId, uint32_t version, ChunkPartType chunkType,
-                        uint16_t *blocks, uint32_t listenerId = 0);
+uint32_t job_get_blocks(JobPool &jobPool, JobPool::JobCallback callback,
+                        GetBlocksHighLevelOp *getBlocksHLO, uint32_t listenerId = 0);
```
Rolled back for the most.

dmga44 commented on src/chunkserver/chunk_high_level_ops.cc:
```diff
@@ -0,0 +1,546 @@
+	chunkId_ = chunkId;
+	chunkVersion_ = chunkVersion;
+	chunkType_ = chunkType;
+
+	getBlocksJobId_ = job_get_blocks(
+	    *workerJobPool(),
+	    [this](uint8_t status, void *entry) { this->getChunkBlocksCallback(status, entry); }, this);
+}
```
We ensure proper lifetime management.

dmga44 commented on src/chunkserver/bgjobs_unittest.cc:
```diff
@@ -190,14 +189,13 @@ TEST_F(JobPoolTest, ChangeCallback) {
-	EXPECT_EQ(counters[kOriginalOpType].load(), 0);
-	EXPECT_EQ(counters[kNewOpType].load(), 1);
+	EXPECT_EQ(counters[kOriginalOpType].load(), 2);
+	EXPECT_EQ(counters[kNewOpType].load(), 0);
```
Rolled back.

dmga44 commented on src/chunkserver/bgjobs.cc:
```diff
@@ -228,11 +228,10 @@ void JobPool::changeCallback(uint32_t jobId, JobCallback callback, void *extra,
 	auto jobIterator = listenerInfo.jobHash.find(jobId);
 	if (jobIterator != listenerInfo.jobHash.end()) {
 		jobIterator->second->callback = std::move(callback);
-		jobIterator->second->extra = extra;
 	}
```
Rolled back.

dmga44 commented on src/chunkserver/bgjobs.h:
```diff
@@ -143,17 +145,15 @@ class JobPool {
-	/// @param extra Additional data to be passed to the new callback.
 	/// @param listenerId The ID of the listener associated with the jobs.
-	void changeCallback(uint32_t jobId, JobCallback callback, void *extra, uint32_t listenerId = 0);
+	void changeCallback(uint32_t jobId, JobCallback callback, uint32_t listenerId = 0);
```
Rolled back.

dmga44 commented on src/chunkserver/chunk_high_level_ops.h:
```diff
@@ -0,0 +1,321 @@
+	/// Type of the chunk associated with the operation
+	ChunkPartType chunkType_ = slice_traits::standard::ChunkPartType();
+	uint16_t pendingDelayedJobs_ = 0;  ///< Number of remaining delayed jobs running
+	bool isChunkOpen_ = 0;             ///< Indicates if the chunk is open.
```
Applied.

dmga44 commented on src/chunkserver/chunkserver_entry.cc:
```diff
@@ -1183,13 +774,8 @@ bool ChunkserverEntry::readHeader(int socket, PacketStruct &packet, uint8_t *hea
-		}
-		// Setting up packet.startPtr is not needed, inputBuffer will be used for reading data
+		writeHLO->prepareForNewWriteData(mustForward, headerBuffer);
+		// Setting up packet.startPtr is not needed, writeHLO will be used for reading data
```
Applied.

dmga44 commented on src/chunkserver/bgjobs.cc:
```diff
@@ -486,17 +482,18 @@ uint32_t job_write(JobPool &jobPool, JobPool::JobCallback callback, void *extra,
-                        uint64_t chunkId, uint32_t version, ChunkPartType chunkType,
-                        uint16_t *blocks, uint32_t listenerId) {
+uint32_t job_get_blocks(JobPool &jobPool, JobPool::JobCallback callback,
+                        GetBlocksHighLevelOp *getBlocksHLO, uint32_t listenerId) {
```
Rolled back.

dmga44 commented on src/chunkserver/chunk_high_level_ops.h:
```diff
@@ -0,0 +1,321 @@
+	/// Type of the chunk associated with the operation
+	ChunkPartType chunkType_ = slice_traits::standard::ChunkPartType();
+	uint16_t pendingDelayedJobs_ = 0;  ///< Number of remaining delayed jobs running
+	bool isChunkOpen_ = 0;             ///< Indicates if the chunk is open.
```
Applied.

dmga44 commented on src/chunkserver/chunk_high_level_ops.h:
```diff
@@ -0,0 +1,310 @@
+
+constexpr uint8_t kSauWriteDataPreffixSize = cltocs::writeData::kPrefixSize;
+// For forwarding: size of SAU_CLTOCS_WRITE_DATA prefix plus the packet header.
+constexpr uint8_t kSauWriteDataPreffixSizeForward =
```
Applied.

dmga44 commented on src/chunkserver/chunk_high_level_ops.cc:
```diff
@@ -0,0 +1,551 @@
+	// - no write data buffers waiting to be enqueued
+	// - no input buffer being filled (all data has been processed)
+	return !isWriteJobBeingProcessed() && partiallyCompletedWrites_.empty() &&
+	       writeDataBuffers_.empty() && inputBuffer_ == nullptr;
```
If is defined, cannot be exactly empty.

dmga44 commented on src/chunkserver/chunk_high_level_ops.h:
```diff
@@ -0,0 +1,310 @@
+	/// @see ChunkserverEntry::readInit
+	void readContinue(uint16_t callMaxParallelHddReadJobs);
+
+	///< Number of blocks to read from the device in one read job.
```
Applied.

lgsilva3087 replied:
///< is used when the comment is in the same line

dmga44 replied:
Thank you, @lgsilva3087, but I think the suggestion was already applied.

dmga44 commented on src/chunkserver/chunk_high_level_ops.h:
```diff
@@ -0,0 +1,310 @@
+/// @brief Base class for high-level chunk operations.
+class HighLevelOp {
+public:
+	HighLevelOp(ChunkserverEntry *parent) : parent_(parent) {};
```
Applied.

dmga44 commented on src/chunkserver/chunk_high_level_ops.cc:
```diff
@@ -0,0 +1,551 @@
+	workerJobPool()->disableJob(writeJobId_);
+	workerJobPool()->changeCallback(
+	    writeJobId_,
+	    [this](uint8_t status, void *entry) { this->delayedCloseCallback(status, entry); },
```
I'll leave it as it is for simplicity.

dmga44 commented on src/chunkserver/chunk_high_level_ops.cc:
```diff
@@ -0,0 +1,551 @@
+	// - no write data buffers waiting to be enqueued
+	// - no input buffer being filled (all data has been processed)
+	return !isWriteJobBeingProcessed() && partiallyCompletedWrites_.empty() &&
+	       writeDataBuffers_.empty() && inputBuffer_ == nullptr;
```
> If the buffer exists but is being updated, treating it as incomplete might be more accurate

I'm actually treating it as incomplete.

dmga44 commented on src/chunkserver/chunk_high_level_ops.h:
```diff
@@ -0,0 +1,310 @@
+	/// Prepares the input buffer for a write operation.
+	void prepareInputBufferForWrite(bool isForward);
+
+	///< Number of blocks to write to the device in one write job.
```
Applied.

dmga44 commented on src/chunkserver/chunk_high_level_ops.cc:
```diff
@@ -0,0 +1,551 @@
+	if (inputBuffer_ != nullptr) {
+		/// Drop the input buffer, it won't be used anymore
+		getWriteInputBufferPool().put(std::move(inputBuffer_));
+	}
+
+	pendingDelayedJobs_++;
+}
+
+void WriteHighLevelOp::cleanup() {
+	while (!writeDataBuffers_.empty()) {
+		getWriteInputBufferPool().put(std::move(writeDataBuffers_.back()));
+		writeDataBuffers_.pop_back();
+	}
+
+	if (isChunkOpen_) {
+		job_close(*workerJobPool(), kEmptyCallback, chunkId_, chunkType_);
+		isChunkOpen_ = false;
+	}
+}
```
Applied.

dmga44 commented on src/chunkserver/chunk_high_level_ops.h:
```diff
@@ -0,0 +1,310 @@
+	/// Performs the delayed close of already discarded read jobs.
+	void delayedClose();
+
+	/// Closes chunk if open.
```
It exactly does what the description says.

dmga44 commented on src/chunkserver/bgjobs.h:
```diff
@@ -51,9 +51,9 @@ class JobPool {
-		Disabled,  /// Job is disabled and will not be processed.
-		Enabled,   /// Job is enabled and ready to be processed.
-		InProgress /// Job is currently being processed.
+		Disabled,   /// Job is disabled and will not be processed.
+		Enabled,    /// Job is enabled and ready to be processed.
+		InProgress  /// Job is currently being processed.
```
Looks like the same to me.

dmga44 commented on src/chunkserver/chunkserver_entry.cc:
```diff
@@ -815,13 +437,11 @@ void ChunkserverEntry::writeEnd(const uint8_t *data, uint32_t length) {
-		 * from a chain !outputPackets.empty() -- there is a status being
+		 * !writeHLO.isCompleted() -- some write data not yet processed
+		 * !outputPackets.empty() -- there is a status being
 		 * send
```
Applied.

dmga44 commented on src/chunkserver/chunk_high_level_ops.h:
```diff
@@ -0,0 +1,310 @@
+class WriteHighLevelOp : public HighLevelOp {
+public:
+	WriteHighLevelOp(ChunkserverEntry *parent, uint16_t maxBlocksPerHddWriteJob)
+	    : HighLevelOp(parent), maxBlocksPerHddWriteJob_(maxBlocksPerHddWriteJob) {};
```
Applied.

dmga44 commented on src/chunkserver/chunk_high_level_ops.h:
```diff
@@ -0,0 +1,310 @@
+	/// Checks if there are jobs that need a delayed close.
+	/// Discards pending read jobs.
+	/// @return True if there are jobs that need a delayed close, false otherwise.
+	bool needsDelayedClose();
```
Applied.

dmga44 commented on src/chunkserver/chunk_high_level_ops.cc:
```diff
@@ -0,0 +1,551 @@
+		/// Drop the input buffer, it won't be used anymore
+		getWriteInputBufferPool().put(std::move(inputBuffer_));
+	}
+
+	pendingDelayedJobs_++;
+}
+
+void WriteHighLevelOp::cleanup() {
+	while (!writeDataBuffers_.empty()) {
+		getWriteInputBufferPool().put(std::move(writeDataBuffers_.back()));
+		writeDataBuffers_.pop_back();
+	}
```
Applied.

lgsilva3087 commented on src/chunkserver/chunk_high_level_ops.h:
```diff
@@ -0,0 +1,310 @@
+class JobPool;
+
+constexpr uint8_t kSauWriteDataPrefixSize = cltocs::writeData::kPrefixSize;
+// For forwarding: size of SAU_CLTOCS_WRITE_DATA prefix plus the packet header.
```
Consider `inline constexpr` in header files.

dmga44 replied:
Applied.

lgsilva3087 reviewed (commented):
Quick initial look.

lgsilva3087 commented on src/chunkserver/chunk_high_level_ops.h:
```diff
@@ -0,0 +1,310 @@
+	inline bool isChunkOpen() const { return isChunkOpen_; }
+
+	/// Returns the ID of the chunk associated with the operation.
+	inline uint64_t chunkId() const { return chunkId_; }
```
The function is implicitly inline here, consider to remove the specifier.

dmga44 replied:
Specifier removed.

lgsilva3087 commented on src/chunkserver/chunk_high_level_ops.h:
```diff
@@ -0,0 +1,310 @@
+	inline void attachBuffer(std::shared_ptr<OutputBuffer> &&buffer);
+
+	/// Creates and attach write status packet on the parent ChunkserverEntry.
+	inline void createAttachedWriteStatus(uint64_t targetChunkId, uint8_t status, uint32_t writeId);
```
These functions are implemented out-of-line (in the .cc) so the compiler most likely will ignore the inline specifier. Please decide which of them could really be inlined, and move the implementation to the header file, which will implicitly inline them.

I see that you access the parent inside some of them, so maybe you can not inline because you need the parent implementation.

dmga44 replied:
At some point of the implementation I did not include the chunkserver_entry.h file to this header, but that's not case now.

I moved their implemention to this file and removed the implicit inline modifier.

lgsilva3087 commented on src/chunkserver/chunk_high_level_ops.h:
```diff
@@ -0,0 +1,310 @@
+	/// Creates and attach write status packet on the parent ChunkserverEntry.
+	inline void createAttachedWriteStatus(uint64_t targetChunkId, uint8_t status, uint32_t writeId);
+
+	uint64_t chunkId_ = 0;       ///< ID of the chunk associated with the operation
+	uint32_t chunkVersion_ = 0;  ///< Version of the chunk associated with the operation
+	/// Type of the chunk associated with the operation
+	ChunkPartType chunkType_ = slice_traits::standard::ChunkPartType();
+	uint16_t pendingDelayedJobs_ = 0;  ///< Number of remaining delayed jobs running
+	bool isChunkOpen_ = false;         ///< Indicates if the chunk is open.
```
Not for this PR maybe, but protected variables are usually avoided in these scenarios to follow better OCP.
Private variables and inline controlled accesors are usually prefferred for better encapsulation.

dmga44 replied:
I actually did not want to use private variables and inline controllers due to extra verbosity. I see these members as intrinsic members of the derived classes so the extra layer seemed weird.

lgsilva3087 commented on src/chunkserver/chunk_high_level_ops.h:
```diff
@@ -0,0 +1,310 @@
+/*
+   Copyright 2013-2015 Skytechnology sp. z o.o.
```
This file is new, maybe this line is not needed.

dmga44 replied:
Yes, but most of its code comes from the chunkserver_entry.cc file (with some few changes), which in has this line.

I have not applied the suggestion.

lgsilva3087 commented on src/chunkserver/chunkserver_entry.cc:
Moving some of the implementation details out of this file create some unused includes, please check if they are really unused:
```cpp
#include <cinttypes>
#include <set>
#include "common/chunk_connector.h"
#include "common/legacy_vector.h"
#include "common/saunafs_version.h"
#include "devtools/request_log.h"
```
Consider to remove them to improve compilation time.

dmga44 replied:
I analyzed it and these were the includes removed:
```cpp
#include <netinet/in.h>
#include <sys/uio.h>
#include <syslog.h>
#include <cinttypes>
#include <cstdlib>
#include <cstring>
#include <ctime>
#include <mutex>
#include <set>
#include "common/legacy_vector.h"
#include "common/saunafs_version.h"
#include "devtools/request_log.h"
```

lgsilva3087 commented on src/chunkserver/chunkserver_entry.h:
```diff
@@ -134,57 +138,19 @@ struct ChunkserverEntry {
+	std::unique_ptr<ReadHighLevelOp> readHLO;
+	/// Get blocks operation related data
+	std::unique_ptr<GetBlocksHighLevelOp> getBlocksHLO;
+
```
The new variables are only used inside ChunkserverEntry, consider to make them private, no need to expose them.

dmga44 replied:
I moved them to private and renamed them, `writeHLO `-> `writeHLO_`.

lgsilva3087 commented on src/chunkserver/chunkserver_entry.cc:
```diff
@@ -69,10 +69,17 @@ static ConnectionPool gForwardConnectionPool;
 constexpr uint32_t kMaxPacketSize = 100000 + SFSBLOCKSIZE;
 constexpr uint8_t kConnectRetries = 10;
```
This comment I know is out of scope, but these should be `static constexpr` unless external linkage is needed, which should not be the case. Maybe we have some ODR violations because the constant is used in other places.

dmga44 replied:
Applied.

lgsilva3087 commented on src/chunkserver/chunk_high_level_ops.h:
```diff
@@ -0,0 +1,310 @@
+
+protected:
+	/// State of the parent ChunkserverEntry.
+	inline ChunkserverEntry::State &state();
```
consider:

```cpp
// .h
ChunkserverEntry::State parentState() const;
void setParentState(ChunkserverEntry::State entryState);

// .cc
void HighLevelOp::setParentState(ChunkserverEntry::State entryState) {
    parent.state = entryState; // Which will work because it is public, but should be a set call too
}
```

Then the lines like:
```cpp
state() = ChunkserverEntry::State::Idle;
```

Could become:
```cpp
setParentState(ChunkserverEntry::State::Idle);
```

dmga44 replied:
I've splitted it into getParentState and setParentState.

lgsilva3087 commented on src/chunkserver/chunk_high_level_ops.h:
```diff
@@ -0,0 +1,310 @@
+	void setup(uint64_t chunkId, uint32_t chunkVersion, ChunkPartType chunkType);
+
+	/// Checks if the get blocks job is currently running.
+	bool isRunning();
```
const

dmga44 replied:
I've added the modifier to this function and some others that could use it.

lgsilva3087 reviewed (commented):
It is looking good, please see my suggestions.

dmga44 committed:
refactor: Apply code review suggestions

Changes:
- applied inline and static modifiers to some constants.
- removed implicit inline modifiers and ignored ones.
- move some HighLevelOp member implementations to the header file to
make them really inline.
- split state member in HighLevelOp into getParentState and
setParentState.
- remove many unused header from chunkserver_entry.cc source file.
- make the high level operations members from the csentry private.

Signed-off-by: Dave <dave@leil.io>

lgsilva3087 reviewed (approved):
Great. LGTM.
